core/src/software/aws/toolkits/core/credentials/sso/DiskCache.kt

```diff
@@ -0,0 +1,117 @@
+
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.fasterxml.jackson.databind.module.SimpleModule
+import com.fasterxml.jackson.databind.util.StdDateFormat
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import software.aws.toolkits.core.utils.deleteIfExists
+import software.aws.toolkits.core.utils.inputStreamIfExists
+import software.aws.toolkits.core.utils.outputStream
+import software.aws.toolkits.core.utils.toHexString
+import software.aws.toolkits.core.utils.tryOrNull
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.security.MessageDigest
+import java.time.Clock
+import java.time.Instant
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatter.ISO_INSTANT
+import java.util.TimeZone
+
+/**
+ * Caches the [AccessToken] to disk to allow it to be re-used with other tools such as the CLI.
+ */
+class DiskCache(
+    private val cacheDir: Path = Paths.get(System.getProperty("user.home"), ".aws", "sso", "cache"),
+    private val clock: Clock = Clock.systemUTC()
+) : SsoCache {
+    private val objectMapper = jacksonObjectMapper().also {
+        val customDateModule = SimpleModule()
+        customDateModule.addDeserializer(Instant::class.java,
+            CliCompatibleInstantDeserializer()
+        )
+
+        it.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+        it.registerModule(JavaTimeModule())
+        it.registerModule(customDateModule) // Override the Instant deserializer with custom one
+        it.dateFormat = StdDateFormat().withTimeZone(TimeZone.getTimeZone(ZoneOffset.UTC))
+    }
+
+    override fun loadClientRegistration(ssoRegion: String): ClientRegistration? {
+        val cacheFile = cacheDir.resolve(clientRegistrationCacheKey(ssoRegion))
+        val inputStream = cacheFile.inputStreamIfExists() ?: return null
+        return tryOrNull {
+            val clientRegistration = objectMapper.readValue<ClientRegistration>(inputStream)
+            if (clientRegistration.expiresAt.isAfter(Instant.now(clock))) {
+                clientRegistration
+            } else {
+                null
+            }
+        }
+    }
+
+    override fun saveClientRegistration(ssoRegion: String, registration: ClientRegistration) {
+        cacheDir.resolve(clientRegistrationCacheKey(ssoRegion)).outputStream().use {
+            objectMapper.writeValue(it, registration)
+        }
+    }
+
+    override fun invalidateClientRegistration(ssoRegion: String) {
+        cacheDir.resolve(clientRegistrationCacheKey(ssoRegion)).deleteIfExists()
+    }
+
+    private fun clientRegistrationCacheKey(ssoRegion: String): String = "aws-toolkit-jetbrains-$ssoRegion.json"
```

Comment:
nit: if we don't use clientRegistrationCacheKey or accessKeyCacheKey anywhere else, why not have this return a path as well to save some repetition of the cacheDir.resolve?

Reply:
Good call, changed